Fixed #13008 -- Changed never_cache to effect actually never caching.

mbertheau opened the pull request:
https://code.djangoproject.com/ticket/13008

MarkusH commented on tests/decorators/tests.py:
```diff
@@ -317,3 +317,13 @@ def a_view(request):
+            return HttpResponse()
+        r = a_view(HttpRequest())
+        self.assertEqual(set(r['Cache-Control'].split(', ')),
+                         set(['max-age=0', 'no-cache', 'no-store', 'must-revalidate']))
```
Please use set literals: `{'max-age=0', 'no-cache', ...}`

MarkusH commented on tests/decorators/tests.py:
```diff
@@ -317,3 +317,13 @@ def a_view(request):
+        def a_view(request):
+            return HttpResponse()
+        r = a_view(HttpRequest())
+        self.assertEqual(set(r['Cache-Control'].split(', ')),
```
Please use a hanging indentation:
```python
self.assertEqual(
    set(...),
    {...},
)

MarkusH commented:
buildbot, ok to test.

mbertheau commented:
Both fixed.

MarkusH commented:
I don't have a strong opinion about it, but the behavior change should probably be noted in the 1.9 release notes: "What's new in Django 1.9" --> "Cache".

mbertheau commented:
Added something to the 1.9 release notes.

claudep commented on django/utils/cache.py:
```diff
@@ -134,6 +134,9 @@ def add_never_cache_headers(response):
     patch_response_headers(response, cache_timeout=-1)
+    patch_cache_control(response, no_cache=True)
+    patch_cache_control(response, no_store=True)
+    patch_cache_control(response, must_revalidate=True)
```
I think that `patch_cache_control` can accept multiple keywords in a single call.

mbertheau replied:
Good point, fixed.

claudep commented:
I'd also complement the docs for https://docs.djangoproject.com/en/dev/ref/utils/#django.utils.cache.add_never_cache_headers

mbertheau commented:
@claudep What information would you add to the docs? It seems pretty complete and now also correct to me.

claudep commented:
I think we should mentioned the exact headers sent. And also add a versionchanged directive to tell the new headers in 1.9.

MarkusH commented:
`never_cache` is not documented at all from what I can see. It's only used in text and code snippets in https://docs.djangoproject.com/en/dev/topics/cache/ . I think we should document it like https://docs.djangoproject.com/en/dev/topics/cache/#django.views.decorators.cache.cache_page

claudep commented:
Markus, see the link in my previous comment, `add_never_cache_headers` is documented.

claudep commented:
Oh, did you mean that `never_cache` should be documented in addition to `add_never_cache_headers`?

MarkusH commented:
Yes, `never_cache` isn't documented.

mbertheau commented:
* Added exact headers sent to `add_never_cache_headers` docs with `versionchanged` tag
* Added docs for `never_cache` with `versionchanged` tag

I'd appreciate input on whether or not to duplicate docs on `never_cache` and `add_never_cache_headers`. Right now it's a mixture - `add_never_cache_headers` has the exact headers and the detailed `versionchanged` note, `never_cache` has a general "doesn't cache" statement and the detailed `versionchanged` note.

timgraham commented on django/views/decorators/cache.py:
```diff
@@ -53,8 +53,8 @@ def never_cache(view_func):
     """
     @wraps(view_func, assigned=available_attrs(view_func))
-    def _wrapped_view_func(request, *args, **kwargs):
+    def _never_cached(request, *args, **kwargs):
```
I'd omit this since it's just a cosmetic change in an otherwise unchanged file.

timgraham commented on docs/ref/utils.txt:
```diff
@@ -65,7 +65,13 @@ need to distinguish caches by the ``Accept-language`` header.
 .. function:: add_never_cache_headers(response)
 
-    Adds headers to a response to indicate that a page should never be cached.
+    Adds "Cache-Control: max-age=0, no-cache, no-store, must-revalidate" header
```
Adds a

timgraham commented on docs/ref/utils.txt:
```diff
@@ -65,7 +65,13 @@ need to distinguish caches by the ``Accept-language`` header.
+    Adds "Cache-Control: max-age=0, no-cache, no-store, must-revalidate" header
+    to a response to indicate that a page should never be cached.
+
+.. versionchanged:: 1.9
```
This annotation and its contents should be indented another level like the others on this page.

timgraham commented on docs/ref/utils.txt:
```diff
@@ -65,7 +65,13 @@ need to distinguish caches by the ``Accept-language`` header.
+.. versionchanged:: 1.9
+
+    Before Django 1.9, "Cache-Control: max-age=0" was sent. This didn't
+    prevent caching reliably in all browsers.
```
chop "reliably" OR phrase it this way: "reliably prevent caching"

timgraham commented on docs/releases/1.9.txt:
```diff
@@ -118,6 +118,9 @@ Cache
 * ``django.core.cache.backends.base.BaseCache`` now has a ``get_or_set()``
   method.
 
+* ``django.views.decorators.cache.never_cache`` now sends more persuasive
```
Can make it a link:
```
:func:`django.views.decorators.cache.never_cache`
```

timgraham commented on docs/releases/1.9.txt:
```diff
@@ -118,6 +118,9 @@ Cache
   method.
 
+* ``django.views.decorators.cache.never_cache`` now sends more persuasive
+  headers to the browser to prevent it from caching (:ticket:`13008`).
```
"... to prevent caching"
no need to include ticket reference in major release notes

timgraham commented on docs/topics/http/decorators.txt:
```diff
@@ -97,3 +97,21 @@ caching based on specific request headers.
+    This decorator adds headers to ensure the response won’t be cached by
+    browsers or other caches.
+
+    .. versionchanged 1.9
```
`.. versionchanged:: 1.9` (missing colons)

timgraham commented on docs/topics/http/decorators.txt:
```diff
@@ -97,3 +97,21 @@ caching based on specific request headers.
+
+    .. versionchanged 1.9
+
+    Before Django 1.9, "Cache-Control: max-age=0" was sent. This didn't
```
indent contents and use whatever wording you choose in the other comment

timgraham commented on docs/topics/http/decorators.txt:
```diff
@@ -97,3 +97,21 @@ caching based on specific request headers.
+
+.. function:: never_cache(view_func)
+
+    This decorator adds headers to ensure the response won’t be cached by
```
no need to use left quote - sphinx will make the correct replacement on the rendered docs

timgraham commented on docs/topics/http/decorators.txt:
```diff
@@ -97,3 +97,21 @@ caching based on specific request headers.
+.. function:: never_cache(view_func)
+
+    This decorator adds headers to ensure the response won’t be cached by
+    browsers or other caches.
```
clarify "other caches"?

mbertheau replied:
I got this from https://github.com/django/django/blob/4870b304f2eabee3ec761f2d94b6251cae418f49/docs/topics/cache.txt#L1213

I can instead use the same text as for `add_never_cache_headers`.

mbertheau replied:
Did that now, let me know :)

timgraham commented on docs/topics/http/decorators.txt:
```diff
@@ -97,3 +97,21 @@ caching based on specific request headers.
+The decorators in :mod:`django.views.decorators.cache` control server and
+client-side caching.
+
+.. function:: never_cache(view_func)
```
This function is mentioned in `topics/cache.txt` and `ref/contrib/admin/index.txt` -- we could make those references links using :func:

mbertheau commented:
All comments addressed. Thanks to everyone for the thorough reviews! :-)

timgraham commented:
merged in 4a438e400b7ce0ab9d0b6876196cbe8d620a4171, thanks!
